### PR TITLE
Make use of the quadtree

### PR DESCRIPTION
**Pull request by Tarcontar**
- [x ] implement proper collider - collider collisions (e.g. ShereCollider - BoxCollider etc)

### CODE REVIEW

File: src/OpenSage.Game/Logic/Object/GameObject.cs [hunk 1]
```diff
@@ -121,9 +120,53 @@ internal void CopyModelConditionFlags(BitArray<ModelConditionFlag> newFlags)
...
+        {
+            Collider.Update(_transform);
+            _gameContext.Quadtree.Update(this);
+        }
+
+        public Vector3 EulerAngles => _transform.EulerAngles;
+        public Vector3 LookDirection => _transform.LookDirection;
+        public Vector3 Translation => _transform.Translation;
+        public Quaternion Rotation => _transform.Rotation;
+        public Matrix4x4 TransformMatrix => _transform.Matrix;
```
Comment (paavohuhtala) [inline src/OpenSage.Game/Logic/Object/GameObject.cs:136]: Does this copy the matrix? Could it return a `ref readonly` instead?

File: src/OpenSage.Game/Logic/Object/GameObject.cs [hunk 2]
```diff
@@ -296,8 +339,8 @@ void AddBehavior(string tag, BehaviorModule behavior)
                 AIUpdate = AddDisposable(objectDefinition.AIUpdate.CreateAIUpdate(this));
                 _tagToModuleLookup.Add(objectDefinition.AIUpdate.Tag, AIUpdate);
             }
-
-            Collider = Collider.Create(objectDefinition, Transform);
+
```
Comment (paavohuhtala) [inline src/OpenSage.Game/Logic/Object/GameObject.cs:342]:
```suggestion

```

**Review by paavohuhtala — CHANGES_REQUESTED** [review]

File: src/OpenSage.Game.Tests/DataStructures/QuadtreeTests.cs [hunk 3]
```diff
@@ -82,38 +85,12 @@ public void InsertAndRemoveNonExistant()
             var item = new MockQuadtreeItem(1, new RectangleF(8, 8, 2, 2));
 
             quadtree.Insert(item);
-            quadtree.Remove(new RectangleF(0, 0, 1, 1));
-
-            Assert.Equal(new[] {item}, quadtree.FindIntersecting(quadtree.Bounds));
-        }
-
-        [Fact]
-        public void InsertAndRemoveManyOverlapping()
```
[the remark this answers is not in the document]
Comment (Tarcontar) [reply src/OpenSage.Game.Tests/DataStructures/QuadtreeTests.cs:91]: cant really remember, so i readded it

File: src/OpenSage.Game/DataStructures/Quadtree.cs [hunk 4]
```diff
@@ -1,111 +1,92 @@
...
-        {
-            var childBounds = GetChild(quad)?.Bounds;
-
-            if (childBounds != null)
-            {
-                return childBounds.Value;
-            }
+            var position = parentBounds.Position;
 
             switch (quad)
```
[the remark this answers is not in the document]
Comment (Tarcontar) [reply src/OpenSage.Game/DataStructures/Quadtree.cs:44]: added comments

File: src/OpenSage.Game/Logic/Object/GameObject.cs (hunk 2 above)
Comment (Tarcontar) [reply src/OpenSage.Game/Logic/Object/GameObject.cs:342]: done

File: src/OpenSage.Game/Logic/Object/GameObject.cs (hunk 1 above)
Comment (Tarcontar) [reply src/OpenSage.Game/Logic/Object/GameObject.cs:136]: ref readonly does not seem to work here

File: src/OpenSage.Game/DataStructures/Quadtree.cs [hunk 5]
```diff
@@ -225,94 +175,53 @@ private void InsertInternal(in RectangleF itemBounds, T item)
...
 
-                    var result = _items[i].Item2;
+        public void Update(in T item)
+        {
+            Remove(item);
+            Insert(item);
+        }
 
-                    _items.RemoveAt(i);
+        public void DebugDraw(DrawingContext2D drawingContext, Camera camera)
```
[the remark this answers is not in the document]
Comment (Tarcontar) [reply src/OpenSage.Game/DataStructures/Quadtree.cs:200]: I agree, but the issue is that i need access to `private Quadtree<T>[] _children;` which is not possible from `DrawingContext2D `etc.

File: src/OpenSage.Game/Gui/DebugUI/DebugOverlay.cs [hunk 6]
```diff
@@ -108,6 +108,8 @@ public void Draw(DrawingContext2D context, Camera camera)
 
             if (ShowColliders)
             {
+                _scene3D.Quadtree.DebugDraw(context, camera);
```
[the remark this answers is not in the document]
Comment (Tarcontar) [reply src/OpenSage.Game/Gui/DebugUI/DebugOverlay.cs:111]: done

File: src/OpenSage.Game/DataStructures/Quadtree.cs [hunk 7]
```diff
@@ -118,62 +99,36 @@ private IEnumerable<T> FindIntersectingInternal(RectangleF itemBounds)
...
+                new Quadtree<T>(Bounds, Quad.UpperRight, depth),
+                new Quadtree<T>(Bounds, Quad.LowerLeft, depth),
+                new Quadtree<T>(Bounds, Quad.LowerRight, depth)
+            };
 
-        public void Insert<A>(A item) where A : T, IHasBounds
-        {
-            Insert(item.Bounds, item);
-        }
+            var oldItems = _items.ToList();
```
[the remark this answers is not in the document]
Comment (Tarcontar) [reply src/OpenSage.Game/DataStructures/Quadtree.cs:122]: done

File: src/OpenSage.Game/DataStructures/Quadtree.cs (hunk 5 above)
[the remark this answers is not in the document]
Comment (paavohuhtala) [reply src/OpenSage.Game/DataStructures/Quadtree.cs:200]: Good point. Then we'll live with this.

**Review by paavohuhtala — APPROVED** [review]

**Pull request merged**